### DISCUSSION TITLE
multiple handlers

### DISCUSSION THREAD

**Discussion by shrugs** [ponder-sh/ponder#387] [Idea / Feature Request]
`Error while building handlers: Cannot add multiple handler functions for event: MyNFT:TransferBatch`

would be nice — currently i'm just using one handler and `await Promise.all([])` to parallelize them, since they're independent. not sure if the expectation is that multiple handlers would be run in parallel or that they'd be run in serial based on order. also gets confusing if you have multiple handlers in different files, since inclusion order would but unintuitive. perhaps instead of a ponder feature this is more of a userspace feature where someone can provide a `compose` function ala

```ts
ponder.on('MyContract:MyEvent', compose([myHandlerOne, myHandlerTwo], { parallel: true }))
```

or something like that

**Comment by typedarray** [comment]
Thanks for opening! Early on, I decided to validate against multiple handlers for the same event because there was no clear way to establish a deterministic order for running them.

Could you share what you mean by "they're independent"? Would be helpful to understand the specific scenario to see how it generalizes.

**Comment by shrugs** [comment]
in my case, in one handler i'm watching nft transfers in order to keep track of who owns what tokens, and in the other handler i'm tabulating volume of transfers over time. obviously the logic for this could live in the same handler, but it felt cleaner to separate them out. 

technically my second handler does reference entities created in the first handler (namely an address' `Account` entity that's created if this is the first time ponder is seeing it) but since the second handler's entities only reference an `Account` by `id`, afaict the `Account` doesn't need to exist yet and there's no race condition related to that reference, so they're happy to be run in parallel.
